crux-core/src/crux/kv/index_store.clj

```diff
@@ -535,6 +536,35 @@
                     (new-prefix-kv-iterator tx-time-mapping-prefix))]
     (latest-completed-tx-i i)))
 
+(defprotocol PThreadManager
+  (snapshot-opened [_ snapshot])
+  (snapshot-closed [_ snapshot]))
+
+(defrecord ThreadManager [^Map snapshot-threads]
+  PThreadManager
+  (snapshot-opened [this snapshot]
+    (locking this
+      (.put snapshot-threads snapshot (Thread/currentThread))))
+
+  (snapshot-closed [this snapshot]
+    (locking this
```

Comment:
I recommend using a ConcurrentHashMap instead of explicit locks.

Reply:
We still need locks here because of the atomic access required to both the map and the semaphore, so I've left it as a map for now?